File: kyulee/programmers-2/타겟넘버.java

public class 타겟넘버 {
    class solution {
        public int solution(int[] numbers, int target) {
            return dfs(numbers, target, 0, 0);
        }

        private int dfs(int[] numbers, int target, int index, int sum) {
            if (index == numbers.length) {
                if (sum == target) {
                    return 1;
                } else {
                    return 0;
                }
            }

            int count = 0;
            count += dfs(numbers, target, index + 1, sum + numbers[index]);
            count += dfs(numbers, target, index + 1, sum - numbers[index]);
            return count;
        }
    }
}

File: sanghoon/programmers-2/타겟 넘버.py
def solution(numbers, target):
    sup = [0]                   # numbers의 0번 인덱스 지정 위해

    for i in numbers:           # numbers의 양,음 모든 경의의 수 sup에 append
        sub = []
        for j in sup:
            sub.append(j + i)
            sub.append(j - i)
        sup = sub
    return sup.count(target)    # target과 동일한 수만 카운트

numbers = [4, 1, 2, 1]
target = 4
solution(numbers, target)	